src/support/hazard.c

```diff
@@ -295,69 +295,81 @@ hazard_get_reference(WT_SESSION_IMPL *session, WT_HAZARD **hazardp, uint32_t *ha
+    }
+
+    for (i = 0; i < hazard_inuse; ++cookie->hp, ++i) {
+        ++cookie->walk_cnt;
+        if (cookie->hp->ref == cookie->search_ref) {
+            WT_STAT_CONN_INCRV(cookie->original_session, cache_hazard_walks, cookie->walk_cnt);
+            if (cookie->session_ret != NULL)
+                *cookie->session_ret = session;
+            *exit_walkp = true;
+            return;
+        }
+    }
+
+    cookie->hp = NULL;
+}
+
 /*
  * __wt_hazard_check --
  *     Return if there's a hazard pointer to the page in the system.
  */
 WT_HAZARD *
 __wt_hazard_check(WT_SESSION_IMPL *session, WT_REF *ref, WT_SESSION_IMPL **sessionp)
 {
-    WT_CONNECTION_IMPL *conn;
-    WT_HAZARD *hp;
-    WT_SESSION_IMPL *s;
-    uint32_t i, j, hazard_inuse, max, session_cnt, walk_cnt;
+    WT_HAZARD_COOKIE cookie;
+
+    WT_CLEAR(cookie);
+    cookie.original_session = session;
+    cookie.session_ret = sessionp;
+    cookie.search_ref = ref;
 
     /* If a file can never be evicted, hazard pointers aren't required. */
     if (F_ISSET(S2BT(session), WT_BTREE_IN_MEMORY))
         return (NULL);
 
-    conn = S2C(session);
-
     WT_STAT_CONN_INCR(session, cache_hazard_checks);
-
     /*
      * Hazard pointer arrays might grow and be freed underneath us; enter the current hazard
      * resource generation for the duration of the walk to ensure that doesn't happen.
      */
     __wt_session_gen_enter(session, WT_GEN_HAZARD);
-
-    /*
-     * No lock is required because the session array is fixed size, but it may contain inactive
-     * entries. We must review any active session that might contain a hazard pointer, so insert a
-     * read barrier after reading the active session count. That way, no matter what sessions come
-     * or go, we'll check the slots for all of the sessions that could have been active when we
-     * started our check.
-     */
-    WT_ORDERED_READ(session_cnt, conn->session_cnt);
-    for (s = conn->sessions, i = max = walk_cnt = 0; i < session_cnt; ++s, ++i) {
-        if (!s->active)
-            continue;
-
-        hazard_get_reference(s, &hp, &hazard_inuse);
-
-        if (hazard_inuse > max) {
-            max = hazard_inuse;
-            WT_STAT_CONN_SET(session, cache_hazard_max, max);
-        }
-
-        for (j = 0; j < hazard_inuse; ++hp, ++j) {
-            ++walk_cnt;
-            if (hp->ref == ref) {
-                WT_STAT_CONN_INCRV(session, cache_hazard_walks, walk_cnt);
-                if (sessionp != NULL)
-                    *sessionp = s;
-                goto done;
-            }
-        }
-    }
-    WT_STAT_CONN_INCRV(session, cache_hazard_walks, walk_cnt);
-    hp = NULL;
-
-done:
+    __wt_session_array_walk(session, __hazard_check_callback, false, &cookie);
     /* Leave the current resource generation. */
     __wt_session_gen_leave(session, WT_GEN_HAZARD);
 
-    return (hp);
+    WT_STAT_CONN_INCRV(session, cache_hazard_walks, cookie.walk_cnt);
```

Comment:
The original code used to skip this line (`goto done`) when it found a hazptr so now we double count the cache walks. Also of lesser note we're now incrementing the stat after calling `gen_leave`. The former `hp = NULL` line is ok as we handle that inside the callback now.

Code fix needed:
```suggestion
    if(cookie.hp == NULL)
        WT_STAT_CONN_INCRV(session, cache_hazard_walks, cookie.walk_cnt);
```